File: tests/test_window_manager_border_calculations.c
/*
 * test_window_manager_border_calculations.c - Unit tests for window manager border calculations
 * Tests the consistency of border/no-border calculations between window manager and TUI
 *
 * This test focuses on ensuring that window height calculations don't include borders
 * to match the TUI's borderless rendering approach.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

// Test framework colors
#define TEST_COLOR_RESET "\033[0m"
#define TEST_COLOR_GREEN "\033[32m"
#define TEST_COLOR_RED "\033[31m"
#define TEST_COLOR_YELLOW "\033[33m"
#define TEST_COLOR_CYAN "\033[36m"

// Test counters
static int tests_run = 0;
static int tests_passed = 0;
static int tests_failed = 0;

// Test utilities
static void print_test_result(const char *test_name, int passed) {
    tests_run++;
    if (passed) {
        tests_passed++;
        printf(TEST_COLOR_GREEN "✓ PASS" TEST_COLOR_RESET " %s\n", test_name);
    } else {
        tests_failed++;
        printf(TEST_COLOR_RED "✗ FAIL" TEST_COLOR_RESET " %s\n", test_name);
    }
}

static void print_summary(void) {
    printf("\n" TEST_COLOR_CYAN "Test Summary:" TEST_COLOR_RESET "\n");
    printf("Tests run: %d\n", tests_run);
    printf(TEST_COLOR_GREEN "Tests passed: %d\n" TEST_COLOR_RESET, tests_passed);
    if (tests_failed > 0) {
        printf(TEST_COLOR_RED "Tests failed: %d\n" TEST_COLOR_RESET, tests_failed);
    } else {
        printf(TEST_COLOR_GREEN "All tests passed!\n" TEST_COLOR_RESET);
    }
}

#define ASSERT(cond) do { \
    if (!(cond)) { \
        printf("%s[FAIL] %s:%d: Assertion failed: %s%s\n", TEST_COLOR_RED, __FILE__, __LINE__, #cond, TEST_COLOR_RESET); \
        return 0; \
    } \
} while(0)

// Simplified window manager config structure
struct TestWindowManagerConfig {
    int min_conv_height;
    int min_input_height;   // Minimum content lines (no borders)
    int max_input_height;   // Maximum content lines (no borders)
    int status_height;
    int padding;
    int conv_h_padding;
    int initial_pad_capacity;
};

// Default configuration values (from window_manager.c line 12-21)
static const struct TestWindowManagerConfig DEFAULT_TEST_CONFIG = {
    .min_conv_height = 5,
    .min_input_height = 2,  // Minimum content lines (no borders)
    .max_input_height = 5,  // Maximum content lines (no borders)
    .status_height = 1,
    .padding = 0,
    .conv_h_padding = 0,
    .initial_pad_capacity = 1000
};

// Test 1: Verify config values don't include borders
static int test_config_no_borders(void) {
    printf("\n%s[TEST] test_config_no_borders%s\n", TEST_COLOR_CYAN, TEST_COLOR_RESET);

    // From window_manager.c line 12-21:
    // .min_input_height = 2,  // Minimum content lines (no borders)
    // .max_input_height = 5,  // Maximum content lines (no borders)

    ASSERT(DEFAULT_TEST_CONFIG.min_input_height == 2);
    ASSERT(DEFAULT_TEST_CONFIG.max_input_height == 5);

    // These values should represent content lines only, not including borders
    // In the old buggy code, these would have been 3 and 6 (content + 2 borders)

    return 1;
}

// Test 2: Verify input resize calculations (no borders)
static int test_input_resize_no_borders(void) {
    printf("\n%s[TEST] test_input_resize_no_borders%s\n", TEST_COLOR_CYAN, TEST_COLOR_RESET);

    // Simulate window_manager_resize_input logic from window_manager.c line 413
    // After fix: int new_height = desired_content_lines;  // content only - no borders

    int desired_content_lines = 3;
    int new_height = desired_content_lines;  // No +2 for borders

    ASSERT(new_height == 3);

    // Clamp to min/max (content lines only)
    if (new_height < DEFAULT_TEST_CONFIG.min_input_height) {
        new_height = DEFAULT_TEST_CONFIG.min_input_height;
    }
    if (new_height > DEFAULT_TEST_CONFIG.max_input_height) {
        new_height = DEFAULT_TEST_CONFIG.max_input_height;
    }

    // Test clamping scenarios
    desired_content_lines = 1;  // Below minimum
    new_height = desired_content_lines;
    if (new_height < DEFAULT_TEST_CONFIG.min_input_height) {
        new_height = DEFAULT_TEST_CONFIG.min_input_height;
    }
    ASSERT(new_height == 2);  // Clamped to min_input_height

    desired_content_lines = 10; // Above maximum
    new_height = desired_content_lines;
    if (new_height > DEFAULT_TEST_CONFIG.max_input_height) {
        new_height = DEFAULT_TEST_CONFIG.max_input_height;
    }
    ASSERT(new_height == 5);  // Clamped to max_input_height

    return 1;
}

// Test 3: Verify layout calculations consistency
static int test_layout_calculations(void) {
    printf("\n%s[TEST] test_layout_calculations%s\n", TEST_COLOR_CYAN, TEST_COLOR_RESET);

    // Simulate screen dimensions
    int screen_height = 24;
    int screen_width = 80;
    (void)screen_width; // Mark as used to avoid warning

    // Window manager configuration
    int input_height = 3;  // Content lines (no borders)
    int status_height = DEFAULT_TEST_CONFIG.status_height;
    int padding = DEFAULT_TEST_CONFIG.padding;

    // Calculate conversation viewport height
    // Formula: screen_height - input_height - status_height - padding
    int conv_viewport_height = screen_height - input_height - status_height - padding;

    ASSERT(conv_viewport_height == 20);  // 24 - 3 - 1 - 0

    // Ensure minimum conversation height
    if (conv_viewport_height < DEFAULT_TEST_CONFIG.min_conv_height) {
        conv_viewport_height = DEFAULT_TEST_CONFIG.min_conv_height;
    }

    // Test with very small screen
    screen_height = 8;
    input_height = 3;
    conv_viewport_height = screen_height - input_height - status_height - padding;

    ASSERT(conv_viewport_height == 4);  // 8 - 3 - 1 - 0

    // Should be clamped to min_conv_height (5)
    if (conv_viewport_height < DEFAULT_TEST_CONFIG.min_conv_height) {
        conv_viewport_height = DEFAULT_TEST_CONFIG.min_conv_height;
    }

    // But we can't have conv_viewport_height > available space
    // In real code, this would adjust input_height downward
    // For test purposes, we'll verify the calculation logic

    return 1;
}

// Test 4: Verify TUI and window manager alignment
static int test_tui_wm_alignment(void) {
    printf("\n%s[TEST] test_tui_wm_alignment%s\n", TEST_COLOR_CYAN, TEST_COLOR_RESET);

    // This test ensures TUI and window manager agree on dimensions

    // TUI constants (from tui.c)
    #define TUI_INPUT_WIN_MIN_HEIGHT 2  // Min height for input window (content lines, no borders)

    // Window manager constants (from window_manager.c)
    #define WM_MIN_INPUT_HEIGHT 2  // Minimum content lines (no borders)

    // They should match!
    ASSERT(TUI_INPUT_WIN_MIN_HEIGHT == WM_MIN_INPUT_HEIGHT);

    // Simulate a scenario
    int screen_height = 24;

    // TUI calculates max height as 20% of screen
    int tui_max_height_percent = 20;
    int tui_calculated_max = (screen_height * tui_max_height_percent) / 100;
    if (tui_calculated_max < TUI_INPUT_WIN_MIN_HEIGHT) {
        tui_calculated_max = TUI_INPUT_WIN_MIN_HEIGHT;
    }

    // 20% of 24 = 4.8, rounded down to 4
    ASSERT(tui_calculated_max == 4);

    // Window manager has fixed max (5)
    int wm_max_height = 5;

    // The effective max should be the smaller of the two
    int effective_max = (tui_calculated_max < wm_max_height) ? tui_calculated_max : wm_max_height;

    ASSERT(effective_max == 4);  // TUI's 4 is smaller than WM's 5

    return 1;
}

// Test 5: Verify the fix for border calculation mismatch
static int test_border_calculation_fix(void) {
    printf("\n%s[TEST] test_border_calculation_fix%s\n", TEST_COLOR_CYAN, TEST_COLOR_RESET);

    // This test demonstrates the bug that was fixed

    // OLD BUGGY CODE (before fix):
    // TUI: Used border calculations (positions +1 for borders)
    // WM: Used content-only calculations (no borders)
    // Result: Mismatch causing scrolling issues

    // NEW FIXED CODE (after fix):
    // Both TUI and WM use content-only calculations (no borders)

    // Simulate old buggy calculation
    int old_tui_cursor_y = 1;  // +1 for border
    int old_tui_cursor_x = 1;  // +1 for border
    int old_wm_input_height = 3;  // Content lines only
    (void)old_tui_cursor_y; // Mark as used to avoid warning
    (void)old_tui_cursor_x; // Mark as used to avoid warning
    (void)old_wm_input_height; // Mark as used to avoid warning

    // In old code, TUI thought window had borders, WM didn't
    // This caused cursor to be positioned incorrectly

    // Simulate new fixed calculation
    int new_tui_cursor_y = 0;  // No border offset
    int new_tui_cursor_x = 0;  // No border offset
    int new_wm_input_height = 3;  // Content lines only

    // Now they match: both use content-only calculations

    // Verify the fix
    ASSERT(new_tui_cursor_y == 0);
    ASSERT(new_tui_cursor_x == 0);

    // The key insight: window manager's input_height should equal
    // the number of lines TUI can draw in (0 to height-1)

    // If WM says input_height = 3, TUI can use rows 0, 1, 2
    for (int row = 0; row < new_wm_input_height; row++) {
        ASSERT(row >= 0 && row < new_wm_input_height);
    }

    return 1;
}

int main(void) {
    printf("%sRunning Window Manager Border Calculation Tests%s\n", TEST_COLOR_CYAN, TEST_COLOR_RESET);

    print_test_result("test_config_no_borders", test_config_no_borders());
    print_test_result("test_input_resize_no_borders", test_input_resize_no_borders());
    print_test_result("test_layout_calculations", test_layout_calculations());
    print_test_result("test_tui_wm_alignment", test_tui_wm_alignment());
    print_test_result("test_border_calculation_fix", test_border_calculation_fix());

    print_summary();

    return (tests_failed == 0) ? 0 : 1;
}

File: zig/tests/test_window_manager_border_calculations.zig
//! tests/test_window_manager_border_calculations.zig
//!   — Zig port of tests/test_window_manager_border_calculations.c
//!
//! Tests the consistency of border/no-border calculations between the
//! window manager and the TUI.  All arithmetic is pure and runs headless.

const std = @import("std");
const wm = @import("../tui/window_manager.zig");

// ---------------------------------------------------------------------------
// Helpers that replicate the C test's inline logic
// ---------------------------------------------------------------------------

/// Clamp `v` to [lo, hi].
fn clamp(v: i32, lo: i32, hi: i32) i32 {
    if (v < lo) return lo;
    if (v > hi) return hi;
    return v;
}

/// Conversation viewport height given a screen and window-manager config.
fn convViewportHeight(cfg: wm.WindowConfig, screen_h: i32, input_h: i32) i32 {
    // padding is 0 in default config
    const raw = screen_h - input_h - cfg.status_height;
    return @max(raw, cfg.min_conv_height);
}

// ---------------------------------------------------------------------------
// Test 1: Config values represent content lines, not content+borders
// ---------------------------------------------------------------------------

test "border calc: default config min_input_height is content-only (no borders)" {
    const cfg = wm.WindowConfig{};
    // Minimum content lines: 2 (not 3 = 2+top/bottom borders)
    try std.testing.expectEqual(@as(i32, 2), cfg.min_input_height);
}

test "border calc: default config max_input_height is content-only (no borders)" {
    const cfg = wm.WindowConfig{};
    // Maximum content lines: 5 (not 6 = 5+borders)
    try std.testing.expectEqual(@as(i32, 5), cfg.max_input_height);
}

// ---------------------------------------------------------------------------
// Test 2: Input resize clamping — no border offset applied
// ---------------------------------------------------------------------------

test "border calc: resize clamp — desired 3 lines is kept as-is" {
    const cfg = wm.WindowConfig{};
    const new_h = clamp(3, cfg.min_input_height, cfg.max_input_height);
    try std.testing.expectEqual(@as(i32, 3), new_h);
}

test "border calc: resize clamp — below min is raised to min_input_height (2)" {
    const cfg = wm.WindowConfig{};
    const new_h = clamp(1, cfg.min_input_height, cfg.max_input_height);
    try std.testing.expectEqual(@as(i32, 2), new_h);
}

test "border calc: resize clamp — above max is capped to max_input_height (5)" {
    const cfg = wm.WindowConfig{};
    const new_h = clamp(10, cfg.min_input_height, cfg.max_input_height);
    try std.testing.expectEqual(@as(i32, 5), new_h);
}

// ---------------------------------------------------------------------------
// Test 3: Layout calculations (screen_height − input − status)
// ---------------------------------------------------------------------------

test "border calc: layout 24-line screen gives conv_viewport 20" {
    const cfg = wm.WindowConfig{};
    // 24 - 3 (input) - 1 (status) - 0 (padding) = 20
    const vh = convViewportHeight(cfg, 24, 3);
    try std.testing.expectEqual(@as(i32, 20), vh);
}

test "border calc: small screen clamps conv_viewport to min_conv_height" {
    const cfg = wm.WindowConfig{};
    // 8 - 3 - 1 = 4, but min_conv_height = 5
    const vh = convViewportHeight(cfg, 8, 3);
    try std.testing.expectEqual(@as(i32, 5), vh);
}

test "border calc: layout invariant — conv_viewport is never below min_conv_height" {
    const cfg = wm.WindowConfig{};
    const screen_heights = [_]i32{ 5, 8, 10, 15, 24, 40, 80 };
    const input_heights = [_]i32{ 2, 3, 4, 5 };

    for (screen_heights) |sh| {
        for (input_heights) |ih| {
            const vh = convViewportHeight(cfg, sh, ih);
            try std.testing.expect(vh >= cfg.min_conv_height);
        }
    }
}

// ---------------------------------------------------------------------------
// Test 4: TUI and window manager alignment on constants
// ---------------------------------------------------------------------------

test "border calc: TUI INPUT_WIN_MIN_HEIGHT matches WM min_input_height" {
    // TUI constant (tui.c: INPUT_WIN_MIN_HEIGHT = 2)
    const tui_min: i32 = 2;
    const cfg = wm.WindowConfig{};
    try std.testing.expectEqual(tui_min, cfg.min_input_height);
}

test "border calc: TUI max input height percentage calculation" {
    const screen_h: i32 = 24;
    const pct: i32 = 20;
    const tui_min: i32 = 2;

    var calculated = (screen_h * pct) / 100;
    if (calculated < tui_min) calculated = tui_min;

    // 20% of 24 = 4 (integer division), which is >= tui_min(2)
    try std.testing.expectEqual(@as(i32, 4), calculated);
}

test "border calc: effective max is min of TUI-computed and WM max_input_height" {
    const screen_h: i32 = 24;
    const tui_computed: i32 = (screen_h * 20) / 100; // = 4
    const cfg = wm.WindowConfig{};

    const effective_max = @min(tui_computed, cfg.max_input_height);
    // tui gives 4, WM gives 5 → effective = 4
    try std.testing.expectEqual(@as(i32, 4), effective_max);
}

// ---------------------------------------------------------------------------
// Test 5: Border calculation fix — cursor starts at (0,0), not (1,1)
// ---------------------------------------------------------------------------

test "border calc: cursor starts at row 0 (no border offset)" {
    const cursor_y: i32 = 0;
    const cursor_x: i32 = 0;
    try std.testing.expectEqual(@as(i32, 0), cursor_y);
    try std.testing.expectEqual(@as(i32, 0), cursor_x);
}

test "border calc: all rows 0..input_height-1 are valid content rows" {
    const input_height: i32 = 3;
    var row: i32 = 0;
    while (row < input_height) : (row += 1) {
        try std.testing.expect(row >= 0 and row < input_height);
    }
}
